release: v0.1.0-alpha.8 (#597)
apollographql/router: CHANGELOG.md

 -->
 
# [v0.1.0-alpha.8] 2022-03-08

## :sparkles: Features

- **Sync and async checkpoints** ([PR #558](https://github.com/apollographql/router/pull/548) and [PR #580](https://github.com/apollographql/router/pull/548))

>   You can now write plugins that can act as check points in services! Checkpoints allow you to make checks, and either return early with a `Response`, or forward a `Request` down the query pipeline.

Author: ```suggestion
    Checkpoints in the request pipeline now allow plugin authors (which includes us!) to check conditions during a request's lifecycle and circumvent further execution if desired.
    
    Using `Step` return types within the checkpoint it's possible to influence what happens (including changing things like the HTTP status code, etc.).  A caching layer, for example, could return `Step::Return(response)` if a cache "hit" occurred and `Step::Continue(request)` (to allow normal processing to continue) in the event of a cache "miss".
    
    These can be either synchronous or asynchronous.  To see examples, see:
    
    - A [synchronous example](https://github.com/apollographql/router/tree/190afe181bf2c50be1761b522fcbdcc82b81d6ca/examples/forbid-anonymous-operations)
    - An [asynchronous example](https://github.com/apollographql/router/tree/190afe181bf2c50be1761b522fcbdcc82b81d6ca/examples/async-allow-client-id)
```

- **Managed Federation no longer requires `router_url`** ([PR #553](https://github.com/apollographql/router/pull/553))

  When using Managed Federation, it is no longer necessary to provide a `routing_url` value.  Instead, the values provided by Studio will be used and the `routing_url` can be used only to override those URLs.

- **Fix plugin ordering** ([PR #559](https://github.com/apollographql/router/issues/559))

  Plugins need to execute in sequence of declaration *except* for certain "core" plugins (e.g., reporting) which must execute early in the plugin sequence to make sure they are in place as soon as possible in the Router lifecycle. This change now ensures that the reporting plugin executes first and that all other plugins are executed in the order of declaration in configuration.

- **Propagate Router operation lifecycle errors** ([PR #537](https://github.com/apollographql/router/issues/537))

  Our recent extension rework was missing a key part: Error propagation and handling! This change makes sure errors that occurred during query planning and query execution will be displayed as GraphQL errors instead of an empty payload.

## :nail_care: Improvements

> - **Introduce Checkpoint and Step** ([PR #558](https://github.com/apollographql/router/pull/558))

Author: ```suggestion
```

> 
>   Layers and extensions writers (which includes us!) now have a mechanism that allows them to let the service orchestrator know whether a service call should be propagated further down the service stack, or it has been fulfilled already and it can bail out. A caching layer, for example, could return `Step::Return(response)` if a cache "hit" occurred and `Step::Continue(request)` in the event of a cache "miss".

Author: ```suggestion
```

    
    These can be either synchronous or asynchronous.  To see examples, see:
    
    - A [synchronous example](https://github.com/apollographql/router/tree/190afe181bf2c50be1761b522fcbdcc82b81d6ca/examples/forbid-anonymous-operations)
    - An [asynchronous example](https://github.com/apollographql/router/tree/190afe181bf2c50be1761b522fcbdcc82b81d6ca/examples/async-allow-client-id)

- **Contracts support** ([PR #573](https://github.com/apollographql/router/pull/573))

  The Apollo Router now supports [Apollo Studio Contracts](https://www.apollographql.com/docs/studio/contracts/)!

- **Add OpenTracing support** ([PR #548](https://github.com/apollographql/router/pull/548))

  OpenTracing support has been added into the reporting plugin.  You're now able to have span propagation (via headers) via two common formats supported by the `opentracing` crate: `zipkin_b3` and `jaeger`.

## :bug: Fixes

> - **Managed Federation no longer requires `router_url`** ([PR #553](https://github.com/apollographql/router/pull/553))
> 
>   When using Managed Federation, it is no longer necessary to provide a `routing_url` value.  Instead, the values provided by Studio will be used and the `routing_url` can be used only to override those URLs.
> 

Author: ```suggestion
- **Configuration no longer requires `router_url`** ([PR #553](https://github.com/apollographql/router/pull/553))

  When using Managed Federation or directly providing a Supergraph file, it is no longer necessary to provide a `routing_url` value.  Instead, the values provided by the Supergraph or Studio will be used and the `routing_url` can be used only to override specific URLs for specific subgraphs.

```

## :bug: Fixes

- **Managed Federation no longer requires `router_url`** ([PR #553](https://github.com/apollographql/router/pull/553))

  When using Managed Federation, it is no longer necessary to provide a `routing_url` value.  Instead, the values provided by Studio will be used and the `routing_url` can be used only to override those URLs.

- **Fix plugin ordering** ([PR #559](https://github.com/apollographql/router/issues/559))

  Plugins need to execute in sequence of declaration *except* for certain "core" plugins (e.g., reporting) which must execute early in the plugin sequence to make sure they are in place as soon as possible in the Router lifecycle. This change now ensures that the reporting plugin executes first and that all other plugins are executed in the order of declaration in configuration.

- **Propagate Router operation lifecycle errors** ([PR #537](https://github.com/apollographql/router/issues/537))

  Our recent extension rework was missing a key part: Error propagation and handling! This change makes sure errors that occurred during query planning and query execution will be displayed as GraphQL errors instead of an empty payload.

> ## :nail_care: Improvements
> 

Author: ```suggestion
```

- **Propagate Router operation lifecycle errors** ([PR #537](https://github.com/apollographql/router/issues/537))

  Our recent extension rework was missing a key part: Error propagation and handling! This change makes sure errors that occurred during query planning and query execution will be displayed as GraphQL errors instead of an empty payload.

## :nail_care: Improvements

- **Introduce Checkpoint and Step** ([PR #558](https://github.com/apollographql/router/pull/558))

  Layers and extensions writers (which includes us!) now have a mechanism that allows them to let the service orchestrator know whether a service call should be propagated further down the service stack, or it has been fulfilled already and it can bail out. A caching layer, for example, could return `Step::Return(response)` if a cache "hit" occurred and `Step::Continue(request)` in the event of a cache "miss".

> # [v0.1.0-alpha.8] 2022-02-25

Author: ```suggestion
# [v0.1.0-alpha.7] 2022-02-25
```
